Make tidy aware of Rust multiline strings

mdboom opened the pull request:
<!-- Please describe your changes on the following line: -->
This makes the internal tidy script properly ignore the contents of Rust multiline strings.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #18551 (github issue number if applicable).

<!-- Either: -->
- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/18553)
<!-- Reviewable:end -->

mdboom committed:
Make tidy aware of Rust multiline strings

As a result of tighter and more correct handling of character
literals, this now catches a few kinds of syntax involving lifetimes
that were previously missed, so those have been updated.

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @cbrewster (or someone else) soon.

highfive commented:
Heads up! This PR modifies the following files:
 * @wafflespeanut: python/tidy/servo_tidy_tests/test_tidy.py, python/tidy/servo_tidy_tests/multiline_string.rs, python/tidy/servo_tidy/tidy.py
 * @edunham: python/tidy/servo_tidy_tests/test_tidy.py, python/tidy/servo_tidy_tests/multiline_string.rs, python/tidy/servo_tidy/tidy.py

wafflespeanut commented:
Thanks for the PR! Looks like this introduces new errors...

    Checking files for tidiness...

    ./components/layout_thread/lib.rs:1686: missing space before -

    ./components/script/dom/bluetooth.rs:250: missing space before }

    ./components/script/dom/bluetooth.rs:250: missing space after {

    ./components/script/dom/headers.rs:341: missing space before -

    ./components/script/dom/htmlscriptelement.rs:466: missing space before -

    ./components/script/dom/htmlscriptelement.rs:471: missing space before -

    ./components/script/dom/htmlscriptelement.rs:496: missing space before }

    ./components/script/dom/htmlscriptelement.rs:496: missing space after {

    ./components/script/dom/htmlscriptelement.rs:500: missing space before }

    ./components/script/dom/htmlscriptelement.rs:500: missing space after {

    ./components/script/dom/htmlscriptelement.rs:515: missing space before }

    ./components/script/dom/htmlscriptelement.rs:515: missing space after {

    ./components/script/dom/htmlscriptelement.rs:604: missing space before =

    ./components/script/dom/htmlscriptelement.rs:616: missing space before =

    ./components/script/dom/htmlscriptelement.rs:617: missing space before /

    ./components/style/values/computed/mod.rs:89: missing space before -

    ./components/style/values/computed/mod.rs:89: missing space before -

mdboom commented:
Yes, thanks.  I missed that there is an `--all` flag on `./mach test-tidy`.  Looking into this now and will update the PR.

mdboom commented:
Updated.  In order to fix the breakage, I tightened up the parsing of character literals to only be things with a single character (including \u{XXXX} forms etc.)  That revealed a small number of places where lifetimes were looking like character literals before and thus tidy issues were incorrectly being ignored, so I've updated those.

I'm sure it goes without saying, but this is really pushing the boundaries of regexes at this point...  🙀

wafflespeanut commented on python/tidy/servo_tidy_tests/test_tidy.py:
```diff
@@ -266,6 +266,11 @@ def test_file_list(self):
 
+    def test_multiline_string(self):
+        errors = tidy.collect_errors_for_files(iterFile('multiline_string.rs'), [], [tidy.check_rust], print_text=True)
+        self.assertEqual(len(list(errors)), 0)
```
Shouldn't we be doing `self.assertNoMoreErrors(errors)` here?

mdboom replied:
Good point.  Updated.

wafflespeanut replied:
Thanks. Please squash your commits, and let's land this :)

wafflespeanut commented:
Please squash the commits.

mdboom commented:
Squashed.

wafflespeanut commented:
Thanks for doing this! @bors-servo r+

bors-servo commented:
:pushpin: Commit c9dafda has been approved by `wafflespeanut`

<!-- @bors-servo r=wafflespeanut c9dafda03a4a143f4626726df7562077a7cdd64f -->

bors-servo commented:
:hourglass: Testing commit c9dafda03a4a143f4626726df7562077a7cdd64f with merge a8a25dac5226a12916c8fe17155d1dbb3b6cb565...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/9025), [arm32](http://build.servo.org/builders/arm32/builds/9262), [arm64](http://build.servo.org/builders/arm64/builds/9297), [linux-dev](http://build.servo.org/builders/linux-dev/builds/9341), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/6146), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/6203), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/9396), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/1842), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/1851), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/6092), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/6155), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/1875), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/1872), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/4657)
Approved by: wafflespeanut
Pushing a8a25dac5226a12916c8fe17155d1dbb3b6cb565 to master...
